### ISSUE TITLE
Externally access Kafka in Kraft mode

### ISSUE THREAD

**Issue by uvinw** [bitnami/bitnami-docker-kafka#265]
### Name and Version

bitnami/kafka:3.2.0

### What steps will reproduce the bug?

I am trying to run a container with Kafka in Kraft mode (without Zookeeper) and allow a producer/consumer duo running on my host machine to connect with it. 

I am following the steps described in the readme, but it is not entirely clear how best to enable external access when running in Kraft mode.

I am creating the container as follows (excuse the syntax as I'm using testcontainers, but the result is similar to running a docker container):

```
//map host port to container port
    addExposedPort(9093)
//env vars
    addEnv("KAFKA_ENABLE_KRAFT", "yes")
    addEnv("KAFKA_CFG_PROCESS_ROLES", "broker,controller")
    addEnv("KAFKA_CFG_CONTROLLER_LISTENER_NAMES", "CONTROLLER")
    addEnv("KAFKA_CFG_LISTENERS", "PLAINTEXT://:9092,CONTROLLER://:9093")
    addEnv("KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP", "CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT")
    addEnv("KAFKA_CFG_CONTROLLER_QUORUM_VOTERS", "1@127.0.0.1:9093")
    addEnv("KAFKA_BROKER_ID", "1")
    addEnv("ALLOW_PLAINTEXT_LISTENER", "yes")

    addEnv("KAFKA_CFG_ADVERTISED_LISTENERS", "PLAINTEXT://127.0.0.1:9092,CONTROLLER://127.0.0.1:9093")
    addEnv("KAFKA_CFG_INTER_BROKER_LISTENER_NAME", "PLAINTEXT")
```

Afterwards, I am trying to create a Kafka client instance using the following prop:
```
props[AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG] = localhost:52402 
//random host port mapped to container port 9093
```

But the container crashes. I tried a few other variants passing a new EXTERNAL port in the KAFKA_CFG_ADVERTISED_LISTENERS but that simply resulted in the connection timing out.

### What is the expected behavior?

Kafka admin, consume and produce to a docker container from the host machine.

### What do you see instead?

The Kafka container logs show the following:

```
kafka 07:59:53.02 

kafka 07:59:53.05 Welcome to the Bitnami kafka container

kafka 07:59:53.07 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-kafka

kafka 07:59:53.09 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-kafka/issues

kafka 07:59:53.11 

kafka 07:59:53.13 INFO  ==> ** Starting Kafka setup **

kafka 07:59:53.78 DEBUG ==> Validating settings in KAFKA_* env vars...

kafka 07:59:53.80 WARN  ==> KAFKA_CFG_BROKER_ID Must match what is set in KAFKA_CFG_CONTROLLER_QUORUM_VOTERS

kafka 07:59:53.82 WARN  ==> KAFKA_CFG_CONTROLLER_QUORUM_VOTERS must match brokers set with KAFKA_CFG_BROKER_ID

kafka 07:59:53.84 WARN  ==> KAFKA_CFG_PROCESS_ROLES must include 'controller' for KRaft

kafka 07:59:53.86 WARN  ==> KAFKA_CFG_LISTENERS must include a listener for CONTROLLER

kafka 07:59:53.92 WARN  ==> You set the environment variable ALLOW_PLAINTEXT_LISTENER=yes. For safety reasons, do not use this flag in a production environment.

kafka 07:59:54.05 INFO  ==> Initializing Kafka...

kafka 07:59:54.12 INFO  ==> No injected configuration files found, creating default config files

kafka 07:59:56.10 INFO  ==> Initializing KRaft...

kafka 07:59:56.12 WARN  ==> KAFKA_KRAFT_CLUSTER_ID not set - If using multiple nodes then you must use the same Cluster ID for each one

kafka 08:00:06.39 INFO  ==> Generated Kafka cluster ID 'AnulgHo3TJiJZ9bQiDVY3Q'

kafka 08:00:06.41 INFO  ==> Formatting storage directories to add metadata...

Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: The advertised.listeners config must not contain KRaft controller listeners from controller.listener.names when process.roles contains the broker role because Kafka clients that send requests via advertised listeners do not send requests to KRaft controllers -- they only send requests to KRaft brokers.

at scala.Predef$.require(Predef.scala:281)

at kafka.server.KafkaConfig.validateAdvertisedListenersDoesNotContainControllerListenersForKRaftBroker$1(KafkaConfig.scala:2048)

at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2114)

at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1997)

at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1470)

at kafka.tools.StorageTool$.$anonfun$main$2(StorageTool.scala:62)

at scala.Option.flatMap(Option.scala:271)

at kafka.tools.StorageTool$.main(StorageTool.scala:62)

at kafka.tools.StorageTool.main(StorageTool.scala)
```

**Comment by halkar** [comment]
I have a similar problem. I replace `CMD` with my own implementation which can for `KAFKA_CFG_ADVERTISED_LISTENERS` to be updated externally after the container starts. Once `KAFKA_CFG_ADVERTISED_LISTENERS` is set I start Kafka in the existing container. But still can't connect.
Getting `dial tcp [::1]:9092: connect: connection refused` where `9092` is the internal port (external container port is different).

**Comment by uvinw** [comment]
@FraPazGal thanks for the hints, but `KAFKA_CFG_ADVERTISED_LISTENERS` is still not working as expected.

Here's what I did: Added the configs are you described above, tried to connect to `props[AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG] = localhost:52402 ` where the port is the one that was mapped to 9094 on the container.

I expected a failure, but instead, I actually made a connection, but to the wrong Kafka instance (I was accidentally running a Kafka Kraft server locally on my host computer). Once I stopped the local Kraft server, my client code failed - so still no luck connecting to the Kraft server on the container.

`2022-06-29 12:08:13.210 [kafka-admin-client-thread | adminclient-1] WARN  o.apache.kafka.clients.NetworkClient - [AdminClient clientId=adminclient-1] Connection to node 1 (localhost/127.0.0.1:9094) could not be established. Broker may not be available.`

The error shows that the client is trying to connect to port 9094, which doesn't exist on the host machine (where my client code runs), rather only exists on the container. So I need a way to tell the container to `advertise` to my client code, which port to use -> and it needs to be one that is is mapped correctly.

**Comment by whithajess** [comment]
@FraPazGal The whole point with `KRaft` is that the broker can be the controller and broker:

![image](https://user-images.githubusercontent.com/2069021/177438709-4cc4dc87-d9c6-4ad9-ae5d-dab1079f9f76.png)

@uvinw The issue Is basically in the error you got:

```
The advertised.listeners config must not contain KRaft controller listeners from controller.listener.names when process.roles contains the broker role because Kafka clients that send requests via advertised listeners do not send requests to KRaft controllers -- they only send requests to KRaft brokers
```

When you do this:

```
addEnv("KAFKA_CFG_ADVERTISED_LISTENERS", "PLAINTEXT://127.0.0.1:9092,CONTROLLER://127.0.0.1:9093")
```

You are exposing the `KRaft Controller` which is essentially performing the role that `zookeeper` used to perform there is no reason to directly connect to this unless you are a `broker` basically remove the controller part of this and access normally your broker on `9092`

**Comment by whithajess** [comment]
> I have a similar problem. I replace `CMD` with my own implementation which can for `KAFKA_CFG_ADVERTISED_LISTENERS` to be updated externally after the container starts. Once `KAFKA_CFG_ADVERTISED_LISTENERS` is set I start Kafka in the existing container. But still can't connect. Getting `dial tcp [::1]:9092: connect: connection refused` where `9092` is the internal port (external container port is different).
> 
> NB It doesn't work if I statically define different internal/external ports `KAFKA_CFG_ADVERTISED_LISTENERS` either.

@halkar note `CMD` in this docker container is non trivial it runs - https://github.com/bitnami/bitnami-docker-kafka/blob/master/3.0/debian-11/rootfs/opt/bitnami/scripts/kafka/run.sh which in turn loads libraries and environments etc. unless you are replacing all of this its not going to work.

**Comment by FraPazGal** [comment]
Hello @whithajess,

Thanks a bunch for sharing your thoughts on this, really appreciated it!! I checked your suggestion and it does solve the issue by modifying the `KAFKA_CFG_ADVERTISED_LISTENERS` env and exposing the `9092` port. Could you confirm this works on your end @uvinw, @halkar?

**Comment by uvinw** [comment]
Thanks @whithajess for your suggestion, but I am still having no luck on my end. My client still tries to connect to the Kafka cluster, which doesn't exist on the clients 9092 port, since the docker mapped port is different (random).

`2022-07-11 10:39:13.516 [kafka-admin-client-thread | adminclient-1] WARN  o.apache.kafka.clients.NetworkClient - [AdminClient clientId=adminclient-1] Connection to node 1 (/127.0.0.1:9092) could not be established. Broker may not be available.`

@FraPazGal excited that you got it working! Would you be able to share all your env vars, along with how you constructed the `bootstrap servers` string?

**Comment by whithajess** [comment]
@uvinw Im not sure around your `testcontainers` syntax but what i use in a `docker-compose.yaml` is as follows:

```
version: '3'
services:
  kafka:
    image: bitnami/kafka:2.8.1
    container_name: local-kafka
    environment:
      - KAFKA_ENABLE_KRAFT=yes
      - KAFKA_CFG_PROCESS_ROLES=broker,controller
      - KAFKA_CFG_CONTROLLER_LISTENER_NAMES=CONTROLLER
      - KAFKA_CFG_LISTENERS=PLAINTEXT://:9092,CONTROLLER://:9093
      - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT
      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://127.0.0.1:9092
      - KAFKA_CFG_BROKER_ID=1
      - KAFKA_CFG_CONTROLLER_QUORUM_VOTERS=1@127.0.0.1:9093
      - ALLOW_PLAINTEXT_LISTENER=yes
      - KAFKA_CFG_AUTO_CREATE_TOPICS_ENABLE=true
      - KAFKA_CFG_NUM_PARTITIONS=3 # Allows up to 3 concurrent workers
    ports:
      - 0.0.0.0:9092:9092
```

with this for boostrap servers: `127.0.0.1:9092`

It sounds like you are having issues with docker networking more than issues with this particular implementation.

Is this from container to container? (Note you can get into specific issues here especially with the likes of Macbook docker installs not honouring `--network=host` and having to use `host.docker.internal` instead of `localhost` or `127.0.0.1`)

Before you get down to actual `kafka` clients and `bootstrap server` settings an easy network connectivity test can be performed via `telnet`

```
$ telnet localhost 9092
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
```

Basically checking if you can connect to the host on that port - I would start here.
